File: go-playground/go-by-examples/mynumbers/two_sum.go
package mynumbers

func TwoSum(nums []int, target int) []int {

	my_map := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		var complement int = target - nums[i]

		map_value_for_component_key, ok := my_map[complement]

		if ok && map_value_for_component_key != i {
			return []int{map_value_for_component_key, i}
		} else {
			my_map[nums[i]] = i
		}
	}

	var emptyArray []int
	return emptyArray
}

File: java-playground/my-implementations/algorithms-and-datastructures/src/main/java/com/my/company/datastructures/arrays/TwoSum.java
package com.my.company.datastructures.arrays;

import org.apache.commons.lang3.tuple.ImmutablePair;

import java.util.HashMap;
import java.util.Map;
public class TwoSum {

  public ImmutablePair twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();

    for (int i = 0; i < nums.length; i++) {
      int complement = target - nums[i];

      if (map.containsKey(complement) && map.get(complement) != i) {
        return new ImmutablePair(map.get(complement), i);
      } else {
        map.put(nums[i], i);
      }
    }
    return null;
  }

  /**
   * The difference between twoSum and twoSum2 is that, with twoSum2, we are running through all the
   * elements of the input array all the time. And that may not be necessary.
   */
  public int[] twoSum2(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();

    for (int i = 0; i < nums.length; i++) {
      map.put(nums[i], i);
    }

    for (int i = 0; i < nums.length; i++) {
      int complement = target - nums[i];
      if (map.containsKey(complement) && map.get(complement) != i) {
        return new int[] {i, map.get(complement)};
      }
    }

    return null;
  }
}
